src/macros.jl

```diff
@@ -237,7 +237,75 @@ macro grad(expr)
         end
     end |> esc
 end
+
+
+"""
+    ReverseDiff.@grad_from_chainrules Base.sin
+
+The `@grad_from_chainrules` macro provides a way to import adjoints defined
+in ChainRules to ReverseDiff.
+"""
+macro grad_from_chainrules(f)
+    @gensym tp output_value output back closure cls_args cls_kwargs
+    return quote
+        $f(args::Vararg{Union{ReverseDiff.TrackedReal, ReverseDiff.TrackedArray}}) = ReverseDiff.track($f, args...)
```

Comment:
I'm not familiar with all internals of ReverseDiff but this seems wrong and I don't think it will work in general. The signature is too wide if the function can only work with numbers and it does not allow to take derivatives with respect to only some of the arguments (this would require definitions for mixed tracked and non-tracked types similar to https://github.com/JuliaDiff/ReverseDiff.jl/blob/01041c8e8237ed42f6414c6fe0f6e6b12162b6ac/src/derivatives/scalars.jl#L12).

BTW it will also define a zero-argument version of the function if I'm not mistaken.

Reply:
> signature is too wide if the function can only work with numbers and it does not allow to take derivatives with respect to only some of the arguments

So, If I understand correctly, I should find all the method definitions of function f, define at least one method for each method I find, only changing the arguments of Real/Array types to its tracked version.

for example, to
```
f(::String, ::Real, ::Array{Real})
```
there should be:

```
f(::String, ::TrackedReal, ::TrackedArray)
f(::String, ::TrackedReal, ::Array)
f(::String, ::Real, ::TrackedArray)
```
three overloads, right?

Reply:
Yes, in this example you would have to define 3 methods. In general, you can't just replace `Real` and `Array{<:Real}`. Eg, scalar rules defined with `@scalar_rule` are defined for `Number` arguments by default. And arrays are also tricky. Eg, there could be two rules for a function with an `AbstractArray{<:Number}` argument where one of them is specialized for `Diagonal`. Then you want to make sure that tracked diagonal matrices end up in the specialized `rrule`. And similarly, if there is only an `rrule` for `Diagonal` you don't want to add a ReverseDiff rule for general tracked arrays.

Reply:
So we can let the user provide a method signature to import the method's rrule, instead of importing all methods of a function by providing only the function name, like:

```julia
@grad_from_chainrules f(::String, ::Number)
```

Reply:
So we can let the user provide a method signature to import the method's rrule, instead of importing all methods of a function by providing only the function name, like:

```julia
@grad_from_chainrules f(::String, ::Number)
```

Reply:
This could be done. There are still problems though for multiple arguments and due to limitations of the pattern matching, in particular for array arguments and if the rule is not defined for `Number` but some (abstract or concrete) subtype. I think a cleaner and safer approach would be to require to specify the exact combination of ReverseDiff types, i.e., e.g.
```julia
@grad_from_chainrules f(::String, ::TrackedReal)
```
and specify the desired combinations of `TrackedReal` and `TrackedArray` etc. manually.